### DISCUSSION TITLE
Improved method for energy calibration

### DISCUSSION THREAD

**Discussion by hagenbuch** [arendst/Tasmota#19474] [Ideas]
Nous A1, Shelly etc. are having a built in chip that is able to do a pretty good AC power (energy consumption) measurement.

However, after flashing Tasmota, one has either to import former calibration data (like for Shelly 3EM) or calibrate the thing by oneself.

**Problem**

Usually, this has two steps:

1. Measuring momentaneus voltage and then setting VoltageSet.
2. Measuring either power or current and then setting PowerSet or CurrentSet.

Thankfully, Tasmota figures out what to do when using PowerSet.

All these measurements are having a time delay between us reading our expensive voltmeters and then imputting the value - while the mains voltage might have already changed.

This lag seems to be unavoidable as far as setting the voltage is concerned - if we don't have a stable grid voltage.

**Now what?**

However, there would be a better way to calibrate current and power:

If we have a ohmic resistor whose value is stable and known and has no drift (I know, we're not using our toaster which will have a huge temperature drift) we could as well input the resistance in Ohms without any time lag:

    OhmSet 500

This would mean we're using a 500 Ohm resistor. If we can rely on voltage being correct (let's assume 230 V for the moment), the resulting current would then be exactly I = U / R = 0,46 A. P = U * I is then 105,8 Watt.

+/- 1% measurement error of the voltage at 230 V and +/- 1% error measuring of 1 A current yields a power error margin of up to 4%!

+/- 1% measurement error of the voltage at 230 V and 0% error measuring the Ohms yields only 2% error margin.

Andreas

**Comment by sfromis** [comment]
Precision power resistors tend to creep up in price, and are not common among people dabbling in IoT, without doing much with electronics projects. And fine tuning should ideally be at more than one level of consumption, for people wanting to fiddle with last digit.

Arguably, typical cheap IoT energy monitoring devices are not really good enough to assume that precise calibration is spot on, anyway.

**Reply by barbudor** [reply]
The resistor would still give you only 1 parameter
Also at the power we are looking at, the heating time and temperature drift would not be much different than those of a toaster or bulb
You wow ld have to know the temp coefficient of the resistor and then have an accurate thermometer to measure it's temperature

Just running in circles....

**Comment by hagenbuch** [comment]
Actually we can heat up the resistor for some minutes and then measure the resistance. Yes, I still would have to measure the voltage but neither power nor current would have to be measured for my suggestion.

The drift of those professional high load resistors I use is less than 100 ppm, that is already pretty good. They are only some Euros. The current in my home wobbles much more, that was my point. I'm not into perfection, just resonable improvements.

**Comment by Jason2866** [comment]
The resistor is more expensive than a good enough 2nd Multimeter.

**Comment by Suitcasechild** [comment]
Maybe here is a solution. this tool is automated and is using the cal method:
At the moment you can use Fluke 45 for automated referencemeasurement or you have a calibrated reference tasmota-plug.
A manual calibration and english version will be implemented soon.

https://github.com/Suitcasechild/AutoCal


<img width="962" height="563" alt="01_first_measurement" src="https://github.com/user-attachments/assets/65bc58bc-0e6a-4708-a78c-8ce3b6545241" />
